en/03_Upgrading/07_Upgrading_Fluent.md

```diff
@@ -0,0 +1,306 @@
+| Content for this record must exist in any locale _or the base (unlocalised) record_. | Content for this record must exist in the current locale _or_ a specific fallback locale, as defined through the Locales admin section. | Content for this record _must_ exist in the current locale to be used. |
+
+[notice]
+Review localisation settings on all of your models and migrate to the correct value. _Do not_ just swap to the new equivalent of your old value unless you explicitly want to keep that behaviour. Instead, consider whether the new functionality warrants a change to a different option.
+[/notice]
+
+### User permissions
+
+Fluent v5 introduced more granular permission management.
+
+|                           | **Fluent 4** | **Fluent 5** | **Fluent 6** |
+|---------------------------|--------------|--------------|--------------|
+| Locale permission support | No           | Yes          | Yes          |
+
+[hint]
+If you're upgrading from Fluent v4, review your requirements for user permissions in the context of locales and update the permissions as needed. For example you may have content authors who should only be allowed to edit content for specific locales.
+[/hint]
+
+### Enhanced CMS UI
+
+Fluent 5 introduced some new configurable UI actions, including a new `Localisation` tab in the edit form for localised models where you see and manage the localisation status of each record. The `Localisation` tab is referred in documentation as the "localisation manager".
+
+|               | `localise_actions_enabled` | `batch_actions_enabled` | `copy_from_locale_enabled` | `copy_to_locale_enabled` |
+|---------------|--------|--------|--------|--------|
+| Default value | `true` | `true` | `true` | `true` |
+| Functionality | Enable localise actions for records which are not available in the current locale (copy to draft, copy & publish and Localise actions) | Enable "Localisation" actions in the edit form. More about this below. | Enable "copy from locale" action in the localisation manager | Enable "copy to locale" action in the localisation manager |
+
+As mentioned in the table above, the `batch_actions_enabled` configuration value controls whether a new "Localisation" action group appears in the edit form. This is not to be confused with "batch actions" menu in the site tree.
+
+![The "Localisation" action group, expanded to show all the actions available in it](../_images/fluent/batch_actions_enabled.png)
+
+[hint]
+Review available CMS UI components and enable those that you need.
+[/hint]
+
+## Fluent state
+
+Fluent state is the mechanism responsible for maintaining the global current locale state. Any time you swap between locales, the fluent state handles this change.
+
+In fluent 5 the way this state affects ORM actions changed.
+
+|              | **Fluent 4** | **Fluent 5** | **Fluent 6** |
+|--------------|--------------|--------------|--------------|
+| Fluent state | Local        | Global       | Global       |
```

Comment:
I'd just like to double check: Was it version 5 where it switched to `Global`? I remembered it being version 6 (but I could certainly be wrong).

Reply:
I think it was `5` as Fluent `6` was mostly focused on versioned history stuff.

Reply:
I took this from a guide Mojmir created originally, but I decided to run the code examples to validate the results.

In fluent 6.x-dev with global state locale "en_US", with no explicit fallbacks defined and default config:
```php
$result = FluentState::singleton()->withState(function (FluentState $state) {
    $state->setLocale('en_NZ');
    return [
        'titles' => Page::get()->column('Title'),
        'list' => Page::get(),
    ];
});

var_dump([
    $result['titles'], // includes titles from the en_NZ locale
    $result['list']->column('Title'), // includes titles from the en_NZ locale
    Page::get()->column('Title') // includes titles from the en_US locale
]);
```

Did I do something wrong there? I would have expected only the top result to use `en_NZ` and the bottom two to use `en_US`, right?

Reply:
Interesting... Yes, I think I would expect the bottom two to use `en_US` if we're stating that there is no more local state.

Reply:
@mfendeksilverstripe do you have any insight here?

Reply:
This is a bit unexpected - maybe Fluent 6 had some changes that I'm not aware of. I remember that Damian was the one who suggested to split some changes off as Fluent 6 from 5 and made the separation. Maybe there were some other changes added into the package along the way.

I think the way forward is to write unit test for all these cases and document the behaviour.

Reply:
If that's the case, are we effectively saying "Whatever the behaviour currently is, that's the correct behaviour"? I'm happy to write those unit tests if that's the case - but I don't have enough background with fluent if the intention is to find out what the behaviour is _and then make decisions about what is expected and what's a bug_

Reply:
I think we need to regroup on that once that once the unit tests are in place but my general recommendation would be to go with `Global` approach as it's a safer and more consistent. I recognise that this may be debatable but preserving local state is additional complexity.

Reply:
It sounds like there's uncertainty as to what the expected behaviour is around this section - I'm going to remove it from the upgrade guide for now (as it appears the behaviour hasn't changed), and I'll open a separate card in `tractorcow/silverstripe-fluent` for working out a) what the current behaviour _is_, b) what it _should be, and c) if the answers to those questions warrant any changes to this guide.

Reply:
Sounds good, thanks 👍

Reply:
https://github.com/tractorcow-farm/silverstripe-fluent/issues/787

Reply:
Removed this section.

Reply:
>    $result['list']->column('Title'), // includes titles from the en_NZ locale

The query state was set to `Fluent.locale = en_NZ` when the query was created (i.e. inside the callback). This state persists after the callback is returned. Read queries are always sticky to the locale when they were created, not when they are executed.

Reply:
> Interesting... Yes, I think I would expect the bottom two to use en_US if we're saying that there is no more local state.

This should apply to fluent 7 on silverstripe 5; I'll leave this to you to figure out as I'm not using ss 5 :)

Reply:
Still local on fluent 6.

```
https://github.com/tractorcow-farm/silverstripe-fluent/blob/6/src/Extension/Traits/FluentObjectTrait.php#L75-L83
```

Reply:
FYI write queries are NOT sticky, and are always in the global locale. :) (I think)